### PR TITLE
Support for Llama4Scout architecture.

### PR DESCRIPTION
**Pull request by gpolovets1**
Adding functional support for Llama4Scout with the following updates: interleaved rope/nope layers, shared experts in transformer blocks, llama3 rope input ordering, temperature tuning, exposing chunked attn in config, MoE fixes and support for router-weighted inputs, and skipping jitting during model init to avoid OOM.

# Tests
[Confirmed logits](https://docs.google.com/document/d/14R1UTUioD871Qd3bG_tek3Yh7-V8DajceJkc1iSzIiw/edit?tab=t.0#heading=h.n761d87663s) match HF on 3 different prompts
[Confirmed perf and accuracy](https://docs.google.com/document/d/1k8gdnKkkhHQRtJ3c-vnusE02DQmbdYw0cc868ccbINs/edit?resourcekey=0-pN7gt_6HlEhNobT7E8fdkQ&tab=t.0) on benchmarks.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_commons/models/jax/recipes/llama4.py [hunk 1]
```diff
@@ -116,107 +137,312 @@ def __init__(self, vllm_config: VllmConfig, rng: PRNGKey, mesh: Mesh):
         self.vllm_config = vllm_config
         self.rng = nnx.Rngs(rng)
         self.mesh = mesh
```
Comment (Chenyaaang) [inline tpu_commons/models/jax/recipes/llama4.py:141]: You can remove this self.mesh and use the one in L159.

File: tpu_commons/models/jax/common/attention/llama4_attention.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        This method computes the attention output by projecting the input `x`
+        to queries, keys, and values, applying RoPE and L2Norm if specified,
+        performing scaled dot-product attention, and projecting the results
+        back to the model dimension.
+        If no RoPE (NoPE) is specified, one can also perform temperature tuning
+        which is useful to combat dilution of attention scores in long-context attention.
+
+        Args:
+            x: The input tensor of shape `(seq_len, d_model)`.
+            op_mode: The operational mode, either 'prefill' or 'generate'.
```
Comment (jrplatin) [inline tpu_commons/models/jax/common/attention/llama4_attention.py]: I think this docstring is out of date

File: tpu_commons/models/jax/common/attention/llama4_attention.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        with jax.named_scope("o_proj"):
+            o_TD = jnp.einsum('TNH,NHD -> TD', outputs_TNH,
+                              self.kernel_o_proj_NHD.value)
+            o_TD = nnx.with_sharding_constraint(
+                o_TD, self.activation_attention_out_td[op_mode])
+        return new_kv_cache, o_TD
+
+    def apply_temperature_tuning(self, md: AttentionMetadata,
+                                 input_arr_TNH: jax.Array) -> jax.Array:
+        """Applies temperature tuning to the input array of shape (T, N, H)."""
```
Comment (jrplatin) [inline tpu_commons/models/jax/common/attention/llama4_attention.py]: Can you add docstring here?

File: tpu_commons/models/jax/common/base.py [hunk 4]
```diff
@@ -167,3 +167,6 @@ def create_kernel_param(self, *args, **kwargs) -> nnx.Param:
     def create_scale_param(self, *args, **kwargs) -> nnx.Param:
         """Creates a scale/gain parameter using the scale_initializer."""
         return self._create_param(self.scale_initializer, *args, **kwargs)
+
+
+# TODO: Add an AdditionalConfig class that documents which parameters can be passed and what they do.
```
Comment (jrplatin) [inline tpu_commons/models/jax/common/base.py:172]: To be addressed here?

Comment (gpolovets1) [reply tpu_commons/models/jax/common/base.py:172]: This will be addressed in a different PR (maybe by Kewei).

File: tpu_commons/models/jax/recipes/llama3.py [hunk 5]
```diff
@@ -200,38 +200,10 @@ def _init_layers(self):
                               sharding_cfg=self.cfg.sharding)
         self.lm_head.generate_kernel(self.rng)
 
-    @classmethod
-    def create_model_with_random_weights(cls, vllm_config: VllmConfig,
```
Comment (bzgoogle) [inline tpu_commons/models/jax/recipes/llama3.py:204]: Hi George, I see the method of creating random weights have been removed. Did you test if the new changes still support random weight?
